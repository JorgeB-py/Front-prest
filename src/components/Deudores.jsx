import React from 'react';
import { Header } from './header';
import { Container, Row, Col, Form } from 'react-bootstrap';
import './styles/Index.css';
import { Footer } from './footer';
import { FormattedMessage } from 'react-intl';
import { useEffect } from 'react';


export default function Index() {
    const [total, setTotal] = React.useState(0);
    const [deudores, setDeudores] = React.useState([]);
    const [filteredDeudores, setFilteredDeudores] = React.useState([]);
    const [searchTerm, setSearchTerm] = React.useState("");

    useEffect(() => {
        const token = localStorage.getItem('token');

        if (token) {
            fetch("http://localhost:3000/deudor", {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${token}`, // Enviar el token en el encabezado
                },
            })
                .then((response) => response.json())
                .then((data) => {
                    setDeudores(data);
                    setFilteredDeudores(data);
                    setTotal(data.reduce((acc, deudor) => acc + deudor.pago_intereses, 0));
                })
                .catch((error) => {
                    console.error('Error al obtener los deudores:', error);
                });
        } else {
            console.error('No hay token disponible');
        }
    }, []);

    // Función para manejar el cambio en el input de búsqueda
    const handleSearchChange = (e) => {
        const value = e.target.value;
        setSearchTerm(value);

        // Filtrar deudores por nombre
        if (value) {
            const filtered = deudores.filter(deudor =>
                deudor.nombrecompleto.toLowerCase().includes(value.toLowerCase())
            );
            setFilteredDeudores(filtered);
        } else {
            setFilteredDeudores(deudores);
        }
    };

    const nav_links = [
        { name: "Deudores", url: "/deudores" },
        { name: "Crear Deudor", url: "/crearcliente" },
        { name: "Consultar Deudor", url: "/consultarcliente" },
    ];
    const nombre_usuario = "Jorge";

    const RenderCards = ({ nombre, fecha }) => {
        return (
            <Col>
                <div className="card card-style" data-testid="deudor-card" style={{ width: "15rem", height: "25rem" }}>
                    <img
                        src="./2148859448.jpg"
                        className="card-img-top"
                        alt="imagen deudor"
                        style={{ height: "200px", objectFit: "cover" }} // Ajusta el tamaño de la imagen
                    />
                    <div className="card-body" style={{ display: "flex", flexDirection: "column", justifyContent: "space-between" }}>
                        <h5 className="card-title">{nombre}</h5>
                        <p className="card-text">{fecha}</p>
                        <a href="/infodeudor" className="btn btn-primary">
                            <FormattedMessage id="app.information" defaultMessage="Information" />
                        </a>
                    </div>
                </div>
            </Col>
        );
    };

    return (
        <>
            <Header nav_links={nav_links} logged={true} usuario={nombre_usuario}></Header>
            <Container style={{ display: "grid", padding: '1rem' }}>
                <h1 style={{ textAlign: 'left' }}><FormattedMessage id="app.welcome" defaultMessage="Welcome" />, Jorge</h1>
                <h3><FormattedMessage id="app.thismonth" defaultMessage="This month you have earned" /></h3>
                <h2 style={{ color: "#004AAC" }}>${total}</h2>
                
                {/* Filtro por nombre */}
                <Form.Control
                    type="text"
                    placeholder="Buscar por nombre"
                    value={searchTerm}
                    onChange={handleSearchChange}
                    style={{ marginTop: '20px' }}
                />
            </Container>
            <Container>
                <Row style={{ padding: "50px" }}>
                    {
                        filteredDeudores.map((deudor, index) => (
                            <RenderCards key={index} nombre={deudor.nombrecompleto} fecha={deudor.fecha}></RenderCards>
                        ))
                    }
                </Row>
            </Container>
            <Footer />
        </>
    );
}
