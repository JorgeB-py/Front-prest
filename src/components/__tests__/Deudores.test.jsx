import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom';
import { IntlProvider } from 'react-intl';
import Deudores from '../Deudores';
import messages_es from '../local/es.json';
import userEvent from '@testing-library/user-event';


const messages = {
  es: messages_es,
};
const language = 'es';

describe('Deudores', () => {
  beforeEach(() => {
    localStorage.setItem('token', 'fake-token');
  
    global.fetch = jest.fn((url, options) => {
      // Validar que se envían los headers correctamente
      expect(options.headers['Authorization']).toBe('Bearer fake-token');
      expect(options.headers['Content-Type']).toBe('application/json');
  
      return Promise.resolve({
        json: () =>
          Promise.resolve([
            { nombrecompleto: 'Deudor 1', fecha: '2024-01-01', pago_intereses: 100 },
            { nombrecompleto: 'Deudor 2', fecha: '2024-01-01', pago_intereses: 200 },
            { nombrecompleto: 'Deudor 3', fecha: '2024-01-01', pago_intereses: 300 },
            { nombrecompleto: 'Deudor 4', fecha: '2024-01-01', pago_intereses: 400 },
            { nombrecompleto: 'Deudor 5', fecha: '2024-01-01', pago_intereses: 500 },
            { nombrecompleto: 'Deudor 6', fecha: '2024-01-01', pago_intereses: 600 },
            { nombrecompleto: 'Deudor 7', fecha: '2024-01-01', pago_intereses: 700 },
            { nombrecompleto: 'Deudor 8', fecha: '2024-01-01', pago_intereses: 800 },
            { nombrecompleto: 'Deudor 9', fecha: '2024-01-01', pago_intereses: 900 },
            { nombrecompleto: 'Deudor 10', fecha: '2024-01-01', pago_intereses: 1000 },
          ]),
      });
    });
  });
  

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Debe renderizar correctamente el componente', () => {
    render(
      <MemoryRouter>
        <IntlProvider locale={language} messages={messages[language]}>
          <Deudores />
        </IntlProvider>
      </MemoryRouter>
    );

    // Verificar textos clave
    expect(screen.getByText("Bienvenido, Jorge")).toBeInTheDocument();
    expect(screen.getByText("Este mes has ganado")).toBeInTheDocument();
  });
  
  test('Debe filtrar deudores correctamente', async () => {
    render(
      <MemoryRouter>
        <IntlProvider locale={language} messages={messages[language]}>
          <Deudores />
        </IntlProvider>
      </MemoryRouter>
    );
  
    // Escribir en el input de búsqueda
    const searchInput = screen.getByPlaceholderText('Buscar por nombre');
    await userEvent.type(searchInput, 'Deudor 1');
  
    // Verificar que solo aparece el deudor filtrado
    await waitFor(() => {
      const cards = screen.getAllByText(/Deudor 1/i);
      expect(cards.length).toBe(1);
      expect(cards[0]).toHaveTextContent('Deudor 1');
  });
  });
  
  

  test('Deben crearse 10 cartas', async () => {
    render(
      <MemoryRouter>
        <IntlProvider locale={language} messages={messages[language]}>
          <Deudores />
        </IntlProvider>
      </MemoryRouter>
    );

    // Esperar que las cartas se carguen
    const cards = await screen.findAllByTestId('deudor-card');

    // Verificar que hay exactamente 10 cartas
    expect(cards.length).toBe(10);
  });
  
  test('Debe aparecer el botón de información y se crean las cartas correctamente', async () => {
    render(
      <MemoryRouter>
        <IntlProvider locale={language} messages={messages[language]}>
          <Deudores />
        </IntlProvider>
      </MemoryRouter>
    );
  
    // Esperar a que se rendericen las cartas
    const cards = await screen.findAllByTestId('deudor-card');
  
    // Verificar que se crean 10 cartas
    expect(cards.length).toBe(10);
  
    // Validar contenido de las cartas
    cards.forEach((card, index) => {
      // Verificar el nombre del deudor
      expect(card).toHaveTextContent(`Deudor ${index + 1}`);
  
      // Verificar la fecha
      expect(card).toHaveTextContent('2024-01-01');
  
      const img = card.querySelector('img');
      expect(img).toBeInTheDocument();
      expect(img).toHaveAttribute('src', './2148859448.jpg');
  
      const infoButton = card.querySelector('a.btn.btn-primary');
      expect(infoButton).toBeInTheDocument();
      expect(infoButton).toHaveTextContent('Información');
    });
  });
  test('Debe calcular y mostrar el total correctamente', async () => {
    render(
      <MemoryRouter>
        <IntlProvider locale={language} messages={messages[language]}>
          <Deudores />
        </IntlProvider>
      </MemoryRouter>
    );
  
    // Esperar que el total sea calculado
    await waitFor(() => {
      expect(screen.getByText('$5500')).toBeInTheDocument(); // Suma de los 10 pagos en el mock
    });
  });
  test('Debe mostrar correctamente los enlaces de navegación', () => {
    render(
      <MemoryRouter>
        <IntlProvider locale={language} messages={messages[language]}>
          <Deudores />
        </IntlProvider>
      </MemoryRouter>
    );
  
    // Verificar que los enlaces de navegación existen
    expect(screen.getByText('Deudores')).toBeInTheDocument();
    expect(screen.getByText('Crear Deudor')).toBeInTheDocument();
    expect(screen.getByText('Consultar Deudor')).toBeInTheDocument();
  });
  test('Cada tarjeta tiene un botón accesible para información', async () => {
    render(
      <MemoryRouter>
        <IntlProvider locale={language} messages={messages[language]}>
          <Deudores />
        </IntlProvider>
      </MemoryRouter>
    );
  
    const buttons = await screen.findAllByRole('link', { name: /Información/i });
    expect(buttons.length).toBe(10); // Un botón por tarjeta
  });
  test('Debe coincidir con el snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <IntlProvider locale={language} messages={messages[language]}>
          <Deudores />
        </IntlProvider>
      </MemoryRouter>
    );
  
    expect(asFragment()).toMatchSnapshot();
  });
        
});
