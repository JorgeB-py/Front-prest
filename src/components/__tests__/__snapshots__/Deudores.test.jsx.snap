// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deudores Debe coincidir con el snapshot 1`] = `
<DocumentFragment>
  <nav
    class="custom-navbar navbar navbar-expand-lg navbar-light"
  >
    <div
      class="container-fluid"
    >
      <a
        class="custom-logo navbar-brand"
        href="/"
      >
        <img
          alt="Logo"
          class="img-fluid"
          height="73"
          src="/Logo_letras.png"
          width="213"
        />
      </a>
      <button
        aria-controls="basic-navbar-nav"
        aria-label="Toggle navigation"
        class="navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="navbar-collapse collapse"
        id="basic-navbar-nav"
      >
        <div
          class="me-auto separation navbar-nav"
        >
          <div
            class="nav-item"
          >
            <a
              class="nav-link custom-links"
              href="/deudores"
            >
              Deudores
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link custom-links"
              href="/crearcliente"
            >
              Crear Deudor
            </a>
          </div>
          <div
            class="nav-item"
          >
            <a
              class="nav-link custom-links"
              href="/consultarcliente"
            >
              Consultar Deudor
            </a>
          </div>
        </div>
        <div
          class="ms-auto navbar-nav"
        >
          <div
            class="nav-item"
          >
            <a
              class="button-empieza-landing btn btn-primary btn-lg"
              href="/perfilPrestamista"
              role="button"
              tabindex="0"
            >
              Jorge
            </a>
          </div>
        </div>
      </div>
    </div>
  </nav>
  <div
    class="container"
    style="display: grid; padding: 1rem;"
  >
    <h1
      style="text-align: left;"
    >
      Bienvenido, Jorge
    </h1>
    <h3>
      Este mes has ganado
    </h3>
    <h2
      style="color: rgb(0, 74, 172);"
    >
      $0
    </h2>
    <input
      class="form-control"
      placeholder="Buscar por nombre"
      style="margin-top: 20px;"
      type="text"
      value=""
    />
  </div>
  <div
    class="container"
  >
    <div
      class="row"
      style="padding: 50px;"
    />
  </div>
  <footer
    class=" text-center text-lg-start footer-color"
  >
    <div
      class="container p-4"
    >
      <img
        alt="Logo"
        class="img-fluid"
        height="73"
        src="/Logo_letras.png"
        style="padding-bottom: 4rem;"
        width="213"
      />
      <div
        class="row"
      >
        <div
          class="col-lg-4 col-md-12 mb-4 mb-md-0"
        >
          <h5
            class="text-uppercase"
          >
            Enlaces
          </h5>
          <ul
            class="list-unstyled mb-0"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link custom-links"
                href="/soluciones"
              >
                Soluciones
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link custom-links"
                href="/nosotros"
              >
                Nosotros
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link custom-links"
                href="/Recursos"
              >
                Recursos
              </a>
            </li>
          </ul>
        </div>
        <div
          class="col-lg-4 col-md-12 mb-4 mb-md-0"
        >
          <h5
            class="text-uppercase"
          >
            Nosotros
          </h5>
          <ul
            class="list-unstyled mb-0"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link custom-links"
                href="/historia"
              >
                La historia de prest
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link custom-links"
                href="/grupo_trabajo"
              >
                Trabajadores de prest
              </a>
            </li>
          </ul>
        </div>
        <div
          class="col-lg-4 col-md-12 mb-4 mb-md-0"
        >
          <h5
            class="text-uppercase"
          >
            Redes Sociales
          </h5>
          <ul>
            <li
              class="nav-item"
            >
              <a
                class="nav-link custom-links"
                href="https://www.tiktok.com/es/"
              >
                TikTok
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link custom-links"
                href="https://www.instagram.com/"
              >
                Instagram
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link custom-links"
                href="https://web.facebook.com/"
              >
                Facebook
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="text-center p-3"
    >
      Â© 2024 Prest. Todos los derechos reservados.
    </div>
  </footer>
</DocumentFragment>
`;
